From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Tue, 14 Jul 2026 00:10:33 -0500
Subject: [PATCH] Fish Handle Redstone Wire Updates

This patch implements Folia's approach to prevent too many chained neighbors race conditions.
This creates a boolean in WorldData to keep track of neighbors per-world.

diff --git a/io/papermc/paper/redstone/RedstoneWireTurbo.java b/io/papermc/paper/redstone/RedstoneWireTurbo.java
index f7820568b78f630e4d234d3dfb36c2167149777a..998722fe12cfb5d9f6b21908abbbbce685ce5307 100644
--- a/io/papermc/paper/redstone/RedstoneWireTurbo.java
+++ b/io/papermc/paper/redstone/RedstoneWireTurbo.java
@@ -827,14 +827,24 @@ public final class RedstoneWireTurbo {
         j = getMaxCurrentStrength(upd, j);
         int l = 0;
 
+        // Fish start - Fix too many chained neighbors
+        final int k;
+        if (net.minecraft.server.MinecraftServer.getServer().fish$worldExecutor.isConcurrentExecutor()) {
+            me.biquaternions.fish.threadedregions.AbstractWorldData fish$worldData = me.biquaternions.fish.threadedregions.TickWorldScheduler.getCurrentRegionizedWorldData();
+            fish$worldData.shouldSignal = false;
+            k = worldIn.getBestNeighborSignal(upd.self);
+            fish$worldData.shouldSignal = true;
+        }
+        else { // Fish end - Fix too many chained neighbors
         wire.shouldSignal = false;
         // Unfortunately, World.isBlockIndirectlyGettingPowered is complicated,
         // and I'm not ready to try to replicate even more functionality from
         // elsewhere in Minecraft into this accelerator.  So sadly, we must
         // suffer the performance hit of this very expensive call.  If there
         // is consistency to what this call returns, we may be able to cache it.
-        final int k = worldIn.getBestNeighborSignal(upd.self);
+        k = worldIn.getBestNeighborSignal(upd.self);
         wire.shouldSignal = true;
+        } // Fish - Fix too many chained neighbors
 
         // The variable 'k' holds the maximum redstone power value of any adjacent blocks.
         // If 'k' has the highest level of all neighbors, then the power level of this
diff --git a/net/minecraft/world/level/block/RedstoneWireBlock.java b/net/minecraft/world/level/block/RedstoneWireBlock.java
index ede64a0afbcdc41e60c3f2f5917c89f09b888433..0e6204cda0e6cdeb157510e1fa8a6647c58b6c7a 100644
--- a/net/minecraft/world/level/block/RedstoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedstoneWireBlock.java
@@ -280,6 +280,7 @@ public class RedstoneWireBlock extends Block {
             if (orientation != null) {
                 source = pos.relative(orientation.getFront().getOpposite());
             }
+            if (worldIn instanceof ServerLevel fish$serverLevel && fish$serverLevel.getServer().fish$worldExecutor.isConcurrentExecutor()) fish$serverLevel.fish$worldData.turbo.updateSurroundingRedstone(worldIn, pos, state, source); else // SparklyPaper - parallel world ticking
             turbo.updateSurroundingRedstone(worldIn, pos, state, source);
             return;
         }
@@ -304,6 +305,7 @@ public class RedstoneWireBlock extends Block {
                 // [Space Walker] suppress shape updates and emit those manually to
                 // bypass the new neighbor update stack.
                 if (level.setBlock(pos, state, Block.UPDATE_KNOWN_SHAPE | Block.UPDATE_CLIENTS)) {
+                    if (level instanceof ServerLevel fish$serverLevel && fish$serverLevel.getServer().fish$worldExecutor.isConcurrentExecutor()) fish$serverLevel.fish$worldData.turbo.updateNeighborShapes(level, pos, state); else // SparklyPaper - parallel world ticking
                     turbo.updateNeighborShapes(level, pos, state);
                 }
             }
@@ -327,6 +329,13 @@ public class RedstoneWireBlock extends Block {
     }
 
     public int getBlockSignal(final Level level, final BlockPos pos) {
+        if (net.minecraft.server.MinecraftServer.getServer().fish$worldExecutor.isConcurrentExecutor()) {
+            me.biquaternions.fish.threadedregions.AbstractWorldData fish$worldData = me.biquaternions.fish.threadedregions.TickWorldScheduler.getCurrentRegionizedWorldData();
+            fish$worldData.shouldSignal = false;
+            int blockSignal = level.getBestNeighborSignal(pos);
+            fish$worldData.shouldSignal = true;
+            return blockSignal;
+        }
         this.shouldSignal = false;
         int blockSignal = level.getBestNeighborSignal(pos);
         this.shouldSignal = true;
@@ -423,12 +432,14 @@ public class RedstoneWireBlock extends Block {
 
     @Override
     protected int getDirectSignal(final BlockState state, final BlockGetter level, final BlockPos pos, final Direction direction) {
+        if (net.minecraft.server.MinecraftServer.getServer().fish$worldExecutor.isConcurrentExecutor()) return !me.biquaternions.fish.threadedregions.TickWorldScheduler.getCurrentRegionizedWorldData().shouldSignal ? Redstone.SIGNAL_MIN : state.getSignal(level, pos, direction);
         return !this.shouldSignal ? Redstone.SIGNAL_MIN : state.getSignal(level, pos, direction);
     }
 
     @Override
     protected int getSignal(final BlockState state, final BlockGetter level, final BlockPos pos, final Direction direction) {
-        if (this.shouldSignal && direction != Direction.DOWN) {
+        boolean shouldSignal0 = net.minecraft.server.MinecraftServer.getServer().fish$worldExecutor.isConcurrentExecutor() ? me.biquaternions.fish.threadedregions.TickWorldScheduler.getCurrentRegionizedWorldData().shouldSignal : this.shouldSignal;
+        if (shouldSignal0 && direction != Direction.DOWN) {
             int power = this.ownSignal(state, level, pos);
             if (power == 0) {
                 return Redstone.SIGNAL_MIN;
@@ -463,6 +474,7 @@ public class RedstoneWireBlock extends Block {
 
     @Override
     protected boolean isSignalSource(final BlockState state) {
+        if (net.minecraft.server.MinecraftServer.getServer().fish$worldExecutor.isConcurrentExecutor()) return me.biquaternions.fish.threadedregions.TickWorldScheduler.getCurrentRegionizedWorldData().shouldSignal;
         return this.shouldSignal;
     }
 
